### ISSUE TITLE
Fix S2259: False positive with switch and pattern matching case blocks

### ISSUE THREAD

**Issue by valhristov** [SonarSource/sonar-dotnet#2240]
### Description

The following FP is raised because we do not put `Not Null` constraint on `o` when performing symbolic execution. We put `Not Null` only on the matched variables.

```
switch (o)
{
    case Foo f:
        ...
        break;

    default:
        return;
}

var s = o.ToString(); // 'o' is null in at least one execution path.
```

The same limitation could appear in other rules too. For example we could get false negatives in S2583 or S2589.

### Related information

* SonarC# Version 7.5+ (after pattern matching support was added to CFG and symbolic execution)
